Public/Src/App/Bxl/AzureDevOpsListener.cs

```diff
@@ -0,0 +1,212 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.ObjectModel;
+using System.Diagnostics.ContractsLight;
+using System.Diagnostics.Tracing;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using BuildXL.Utilities.Collections;
+using BuildXL.Utilities.Instrumentation.Common;
+using BuildXL.Utilities.Tracing;
+using BuildXL.ViewModel;
+
+namespace BuildXL
+{
+    /// <summary>
+    /// This event listener should only be hooked up with AzureDevOps optimzied UI is requested
+    /// It tracks data that is not easily captured at the site where the information is emitted because
+    /// it is too hard to track through. It is most ideal to capture data at the site of collections.
+    /// But as an example the cache miss analyzer is many levels of indirection deep and passing
+    /// the BuildSummary model down would be way to cumbersome.
+    /// </summary>
+    public sealed class AzureDevOpsListener : FormattingEventListener
+    {
+        private static readonly char[] s_newLineCharArray = Environment.NewLine.ToCharArray();
+
+        private readonly IConsole m_console;
+
+        /// <summary>
+        /// The last reported percentage. To avoid double reporting the same percentage over and over
+        /// </summary>
+        private int m_lastReportedProgress = -1;
+
+        private readonly BuildViewModel m_buildViewModel;
+
+        /// <nodoc />
+        public AzureDevOpsListener(
+            Events eventSource,
+            IConsole console,
+            DateTime baseTime,
+            BuildViewModel buildViewModel)
+            : base(eventSource, baseTime, null, EventLevel.Verbose, false, TimeDisplay.Seconds)
+        {
+            Contract.Requires(eventSource != null);

```

Comment:
> Contract.Requires(eventSource != null); [](start = 11, length = 40)

nit: I believe the base class checks this condition; not sure if we need it here as well